# 1.0 Data Cleaning Walkthrough: Combining the Data

## 1.1 Introduction

In the last mission, we began investigating possible relationships between **SAT scores** and **demographic factors**. In order to do this, we acquired several data sets about [New York City public schools](https://data.cityofnewyork.us/data?cat=education). We manipulated these data sets, and found that we could combine them all using the **DBN** column. All of the data sets are currently stored as **keys** in the **data** dictionary. Each individual data set is a pandas dataframe.

In this section, **we'll clean the data a bit more**, then **combine** it. Finally, we'll **compute correlations** and perform some analysis.

The first thing we'll need to do in preparation for the merge is condense some of the data sets. In the last section, we noticed that the values in the **DBN** column were unique in the **sat_results** data set. Other data sets like **class_size** had duplicate **DBN** values, however.

We'll need to condense these data sets so that each value in the **DBN** column is unique. If not, we'll run into issues when it comes time to combine the data sets.

While the main data set we want to analyze, **sat_results**, has unique **DBN** values for every high school in New York City, other data sets aren't as clean. A single row in the **sat_results** data set may match multiple rows in the **class_size** data set, for example. This situation will create problems, because we don't know which of the multiple entries in the **class_size** data set we should combine with the single matching entry in **sat_results**. Here's a diagram that illustrates the problem:


<left><img width="400" src="https://drive.google.com/uc?export=view&id=1deYm5RdQXO2xMX6dUgHLvqDEWipk3axq"></left>

In the diagram above, we can't just combine the rows from both data sets because there are several cases where multiple rows in **class_size** match a single row in **sat_results.**

To resolve this issue, we'll condense the **class_size**, **graduation**, and **demographics** data sets so that each **DBN** is unique.

## 1.2 Condensing the Class Size Data Set

The first data set that we'll condense is **class_size**. The first few rows of **class_size** look like this:

|__| CSD | BOROUGH | SCHOOL CODE | SCHOOL NAME               | GRADE | PROGRAM TYPE | CORE SUBJECT (MS CORE and 9-12 ONLY) | CORE COURSE (MS CORE and 9-12 ONLY) | SERVICE CATEGORY(K-9* ONLY) | NUMBER OF STUDENTS / SEATS FILLED | NUMBER OF SECTIONS |
|---|-----|---------|-------------|---------------------------|-------|--------------|--------------------------------------|-------------------------------------|-----------------------------|-----------------------------------|--------------------|
| 0 | 1   | M       | M015        | P.S. 015 Roberto Clemente | 0K    | GEN ED       | -                                    | -                                   | -                           | 19.0                              | 1.0                |
| 1 | 1   | M       | M015        | P.S. 015 Roberto Clemente | 0K    | CTT          | -                                    | -                                   | -                           | 21.0                              | 1.0                |
| 2 | 1   | M       | M015        | P.S. 015 Roberto Clemente | 01    | GEN ED       | -                                    | -                                   | -                           | 17.0                              | 1.0                |
| 3 | 1   | M       | M015        | P.S. 015 Roberto Clemente | 01    | CTT          | -                                    | -                                   | -                           | 17.0                              | 1.0                |
| 4 | 1   | M       | M015        | P.S. 015 Roberto Clemente | 02    | GEN ED       | -                                    | -                                   | -                           | 15.0                              | 1.0                |

As you can see, the first few rows all pertain to the same school, which is why the **DBN** appears more than once. It looks like each school has multiple values for **GRADE**, **PROGRAM TYPE**, **CORE SUBJECT (MS CORE and 9-12 ONLY)**, and **CORE COURSE (MS CORE and 9-12 ONLY)**.

If we look at the unique values for **GRADE**, we get the following:

```python
array(['0K', '01', '02', '03', '04', '05', '0K-09', nan, '06', '07', '08',
       'MS Core', '09-12', '09'], dtype=object)
```

Because we're dealing with high schools, we're only concerned with grades 9 through 12. That means we only want to pick rows where the value in the **GRADE** column is **09-12**.

If we look at the unique values for **PROGRAM TYPE**, we get the following:

```python
array(['GEN ED', 'CTT', 'SPEC ED', nan, 'G&T'], dtype=object)
```

Each school can have multiple program types. Because **GEN ED** is the largest category by far, let's only select rows where **PROGRAM TYPE** is **GEN ED**.



## 1.3 Condensing the Class Size Data Set

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>


- Create a new variable called **class_size**, and assign the value of **data["class_size"]** to it.
- Filter **class_size** so the **GRADE** column only contains the value **09-12.** Note that the name of the **GRADE** column has a space at the end; you'll generate an error if you don't include it.
- Filter **lass_size** so that the **PROGRAM TYPE** column only contains the value **GEN ED.**
- Display the first five rows of **class_size** to verify.

In [0]:
# put your code here
import pandas as pd
data_files = [
    "ap_2010.csv",
    "class_size.csv",
    "demographics.csv",
    "graduation.csv",
    "hs_directory.csv",
    "sat_results.csv"
]
data = {}
for f in data_files:
  d = pd.read_csv("{0}".format(f))
  key_name = f.replace(".csv", "")
  data[key_name] = d
  print(d.shape)

(258, 5)
(27611, 16)
(10075, 38)
(25096, 23)
(435, 64)
(478, 6)


In [0]:
all_survey = pd.read_csv("survey_all.txt", delimiter="\t", encoding="windows-1252")
d75_survey = pd.read_csv("survey_d75.txt", delimiter="\t", encoding="windows-1252")
survey = pd.concat([all_survey,d75_survey], axis=0)
survey["DBN"] = survey["dbn"]
survey = survey.loc[:,["DBN", "rr_s", "rr_t", "rr_p", "N_s", "N_t", "N_p", "saf_p_11", "com_p_11", 
 "eng_p_11", "aca_p_11", "saf_t_11", "com_t_11", "eng_t_11", 
 "aca_t_11", "saf_s_11", "com_s_11", "eng_s_11", "aca_s_11",
 "saf_tot_11", "com_tot_11", "eng_tot_11", "aca_tot_11"]]
survey.head()

DBN  rr_s  rr_t  rr_p    N_s   N_t    N_p  saf_p_11  com_p_11  eng_p_11  \
0  01M015   NaN    88    60    NaN  22.0   90.0       8.5       7.6       7.5   
1  01M019   NaN   100    60    NaN  34.0  161.0       8.4       7.6       7.6   
2  01M020   NaN    88    73    NaN  42.0  367.0       8.9       8.3       8.3   
3  01M034  89.0    73    50  145.0  29.0  151.0       8.8       8.2       8.0   
4  01M063   NaN   100    60    NaN  23.0   90.0       8.7       7.9       8.1   

      ...      eng_t_11  aca_t_11  saf_s_11  com_s_11  eng_s_11  aca_s_11  \
0     ...           7.6       7.9       NaN       NaN       NaN       NaN   
1     ...           8.9       9.1       NaN       NaN       NaN       NaN   
2     ...           6.8       7.5       NaN       NaN       NaN       NaN   
3     ...           6.8       7.8       6.2       5.9       6.5       7.4   
4     ...           7.8       8.1       NaN       NaN       NaN       NaN   

   saf_tot_11  com_tot_11  eng_tot_11  aca_tot_11  
0         8.0         7.7         7.5         7.9  
1         8.5         8.1         8.2         8.4  
2         8.2         7.3         7.5         8.0  
3         7.3         6.7         7.1         7.9  
4         8.5         7.6         7.9         8.0  

[5 rows x 23 columns]

In [0]:
data["survey"] = survey
data["survey"].shape

(1702, 23)

In [0]:
class_size = data["class_size"]
class_size = class_size[class_size["GRADE "]=="09-12"]
class_size = class_size[class_size["PROGRAM TYPE"] == "GEN ED"]
class_size.head()

CSD BOROUGH SCHOOL CODE                                    SCHOOL NAME  \
225    1       M        M292  Henry Street School for International Studies   
226    1       M        M292  Henry Street School for International Studies   
227    1       M        M292  Henry Street School for International Studies   
228    1       M        M292  Henry Street School for International Studies   
229    1       M        M292  Henry Street School for International Studies   

    GRADE  PROGRAM TYPE CORE SUBJECT (MS CORE and 9-12 ONLY)  \
225  09-12       GEN ED                              ENGLISH   
226  09-12       GEN ED                              ENGLISH   
227  09-12       GEN ED                              ENGLISH   
228  09-12       GEN ED                              ENGLISH   
229  09-12       GEN ED                                 MATH   

    CORE COURSE (MS CORE and 9-12 ONLY) SERVICE CATEGORY(K-9* ONLY)  \
225                           English 9                           -   
226                          English 10                           -   
227                          English 11                           -   
228                          English 12                           -   
229                  Integrated Algebra                           -   

     NUMBER OF STUDENTS / SEATS FILLED  NUMBER OF SECTIONS  \
225                               63.0                 3.0   
226                               79.0                 3.0   
227                               38.0                 2.0   
228                               69.0                 3.0   
229                               53.0                 3.0   

     AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  SIZE OF LARGEST CLASS  \
225                21.0                    19.0                   25.0   
226                26.3                    24.0                   31.0   
227                19.0                    16.0                   22.0   
228                23.0                    13.0                   30.0   
229                17.7                    16.0                   21.0   

    DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  
225       STARS                             NaN  
226       STARS                             NaN  
227       STARS                             NaN  
228       STARS                             NaN  
229       STARS                             NaN

## 1.4 Computing Average Class Sizes

As we saw when we displayed **class_size** on the last screen, **DBN** still isn't completely unique. This is due to the **CORE COURSE (MS CORE and 9-12 ONLY)** and **CORE SUBJECT (MS CORE and 9-12 ONLY)** columns.

**CORE COURSE (MS CORE and 9-12 ONLY)** and **CORE SUBJECT (MS CORE and 9-12 ONLY)** seem to pertain to different kinds of classes. For example, here are the unique values for **CORE SUBJECT (MS CORE and 9-12 ONLY)**:

```python
array(['ENGLISH', 'MATH', 'SCIENCE', 'SOCIAL STUDIES'], dtype=object)
```

This column only seems to include certain subjects. We want our class size data to include every single class a school offers -- not just a subset of them. What we can do is take the average across all of the classes a school offers. This will give us unique **DBN** values, while also incorporating as much data as possible into the average.

Fortunately, we can use the [pandas.DataFrame.groupby()](http://pandas.pydata.org/pandas-docs/stable/groupby.html) method to help us with this. The **DataFrame.groupby()** method will split a dataframe up into unique groups, based on a given column. We can then use the **agg()** method on the resulting **pandas.core.groupby** object to find the **mean** of each column.

Let's say we have this data set:

<left><img width="500" src="https://drive.google.com/uc?export=view&id=1sJjENlTRR56RwYzBmmsU8aIMELgjx8zg"></left>

Using the **groupby()** method, we'll split this dataframe into four separate groups -- one with the **DBN 01M292**, one with the **DBN 01M332**, one with the **DBN 01M378**, and one with the **DBN 01M448**:

<left><img width="500" src="https://drive.google.com/uc?export=view&id=1y9imbMLKRDI50wQqPn7P6TAd6MfCL4Nq"></left>

<left><img width="500" src="https://drive.google.com/uc?export=view&id=1FitnyClxHDQLnoAB3jR7YI_jEPZZhkco"></left>

Then, we can compute the averages for the **AVERAGE CLASS SIZE** column in each of the four groups using the **agg()** method:

<left><img width="200" src="https://drive.google.com/uc?export=view&id=1gHVZixGOuGYYON_zU0OUPTJcC9Q_mKeV"></left>

After we group a dataframe and aggregate data based on it, the column we performed the grouping on (in this case **DBN**) will become the index, and will no longer appear as a column in the data itself. To undo this change and keep **DBN** as a column, we'll need to use [pandas.DataFrame.reset_index()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.reset_index.html). This method will reset the index to a list of integers and make **DBN** a column again.

## 1.5 Computing Average Class Sizes

- Find the average values for each column associated with each **DBN** in **class_size**.
    - Use the [pandas.DataFrame.groupby()](http://pandas.pydata.org/pandas-docs/stable/groupby.html) method to group **class_size** by **DBN**.
    - Use the [agg()](http://pandas.pydata.org/pandas-docs/stable/groupby.html#aggregation) method on the resulting **pandas.core.groupby** object, along with the **numpy.mean()** function as an argument, to calculate the average of each group.
    - Assign the result back to **class_size**.
- Reset the index to make **DBN** a column again.
    - Use the [pandas.DataFrame.reset_index()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.reset_index.html) method, along with the keyword argument **inplace=True**.
- Assign **class_size** back to the **class_size** key of the **data** dictionary.
- Display the first few rows of **data["class_size"]** to verify that everything went okay.

In [0]:
# put your code here
def pad_csd(num):
    string_representation = str(num)
    if len(string_representation) > 1:
        return string_representation
    else:
        return string_representation.zfill(2)
      


class_size["CSD"]= class_size["CSD"].apply(pad_csd)

In [0]:
class_size["DBN"] = class_size["CSD"] + class_size["SCHOOL CODE"]
class_size.head()

CSD BOROUGH SCHOOL CODE                                    SCHOOL NAME  \
225  01       M        M292  Henry Street School for International Studies   
226  01       M        M292  Henry Street School for International Studies   
227  01       M        M292  Henry Street School for International Studies   
228  01       M        M292  Henry Street School for International Studies   
229  01       M        M292  Henry Street School for International Studies   

    GRADE  PROGRAM TYPE CORE SUBJECT (MS CORE and 9-12 ONLY)  \
225  09-12       GEN ED                              ENGLISH   
226  09-12       GEN ED                              ENGLISH   
227  09-12       GEN ED                              ENGLISH   
228  09-12       GEN ED                              ENGLISH   
229  09-12       GEN ED                                 MATH   

    CORE COURSE (MS CORE and 9-12 ONLY) SERVICE CATEGORY(K-9* ONLY)  \
225                           English 9                           -   
226                          English 10                           -   
227                          English 11                           -   
228                          English 12                           -   
229                  Integrated Algebra                           -   

     NUMBER OF STUDENTS / SEATS FILLED  NUMBER OF SECTIONS  \
225                               63.0                 3.0   
226                               79.0                 3.0   
227                               38.0                 2.0   
228                               69.0                 3.0   
229                               53.0                 3.0   

     AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  SIZE OF LARGEST CLASS  \
225                21.0                    19.0                   25.0   
226                26.3                    24.0                   31.0   
227                19.0                    16.0                   22.0   
228                23.0                    13.0                   30.0   
229                17.7                    16.0                   21.0   

    DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO     DBN  
225       STARS                             NaN  01M292  
226       STARS                             NaN  01M292  
227       STARS                             NaN  01M292  
228       STARS                             NaN  01M292  
229       STARS                             NaN  01M292

In [0]:
import numpy as np
grouped = class_size.groupby("DBN")
class_size = grouped.agg(np.mean)

In [0]:
class_size.reset_index("DBN",inplace=True)

In [0]:
data["class_size"] = class_size
data["class_size"].head()

DBN  NUMBER OF STUDENTS / SEATS FILLED  NUMBER OF SECTIONS  \
0  01M292                            88.0000            4.000000   
1  01M332                            46.0000            2.000000   
2  01M378                            33.0000            1.000000   
3  01M448                           105.6875            4.750000   
4  01M450                            57.6000            2.733333   

   AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  SIZE OF LARGEST CLASS  \
0           22.564286                   18.50              26.571429   
1           22.000000                   21.00              23.500000   
2           33.000000                   33.00              33.000000   
3           22.231250                   18.25              27.062500   
4           21.200000                   19.40              22.866667   

   SCHOOLWIDE PUPIL-TEACHER RATIO  
0                             NaN  
1                             NaN  
2                             NaN  
3                             NaN  
4                             NaN

## 1.6 Condensing the Demographics Data Set

Now that we've finished condensing **class_size**, let's condense **demographics**. The first few rows look like this:

| _| DBN    | Name                      | schoolyear | fl_percent | frl_percent | total_enrollment | prek | k  | grade1 | grade2 |
|---|--------|---------------------------|------------|------------|-------------|------------------|------|----|--------|--------|
| 0 | 01M015 | P.S. 015 ROBERTO CLEMENTE | 20052006   | 89.4       | NaN         | 281              | 15   | 36 | 40     | 33     |
| 1 | 01M015 | P.S. 015 ROBERTO CLEMENTE | 20062007   | 89.4       | NaN         | 243              | 15   | 29 | 39     | 38     |
| 2 | 01M015 | P.S. 015 ROBERTO CLEMENTE | 20072008   | 89.4       | NaN         | 261              | 18   | 43 | 39     | 36     |
| 3 | 01M015 | P.S. 015 ROBERTO CLEMENTE | 20082009   | 89.4       | NaN         | 252              | 17   | 37 | 44     | 32     |
| 4 | 01M015 | P.S. 015 ROBERTO CLEMENTE | 20092010   |  _          | 96.5        | 208              | 16   | 40 | 28     | 32     |

In this case, the only column that prevents a given **DBN** from being unique is **schoolyear**. We only want to select rows where schoolyear is **20112012**. This will give us the most recent year of data, and also match our SAT results data.

## 1.7 Condensing the Demographics Data Set

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Filter **demographics**, only selecting rows in **data["demographics"]** where **schoolyear** is **20112012.**
    - **schoolyear** is actually an integer, so be careful about how you perform your comparison.
- Display the first few rows of **data["demographics"]** to verify that the filtering worked.

In [0]:
# put your code here
data["demographics"] = data["demographics"].loc[data["demographics"]["schoolyear"]==20112012]
data["demographics"].head()

DBN                                              Name  schoolyear  \
6   01M015  P.S. 015 ROBERTO CLEMENTE                           20112012   
13  01M019  P.S. 019 ASHER LEVY                                 20112012   
20  01M020  PS 020 ANNA SILVER                                  20112012   
27  01M034  PS 034 FRANKLIN D ROOSEVELT                         20112012   
35  01M063  PS 063 WILLIAM MCKINLEY                             20112012   

   fl_percent  frl_percent  total_enrollment prek    k grade1 grade2  \
6         NaN         89.4               189   13   31     35     28   
13        NaN         61.5               328   32   46     52     54   
20        NaN         92.5               626   52  102    121     87   
27        NaN         99.7               401   14   34     38     36   
35        NaN         78.9               176   18   20     30     21   

      ...     black_num black_per hispanic_num hispanic_per white_num  \
6     ...            63      33.3          109         57.7         4   
13    ...            81      24.7          158         48.2        28   
20    ...            55       8.8          357         57.0        16   
27    ...            90      22.4          275         68.6         8   
35    ...            41      23.3          110         62.5        15   

   white_per male_num male_per female_num female_per  
6        2.1     97.0     51.3       92.0       48.7  
13       8.5    147.0     44.8      181.0       55.2  
20       2.6    330.0     52.7      296.0       47.3  
27       2.0    204.0     50.9      197.0       49.1  
35       8.5     97.0     55.1       79.0       44.9  

[5 rows x 38 columns]

## 1.8 Condensing the Graduation Data Set

Finally, we'll need to condense the **graduation** data set. Here are the first few rows:

| _ | Demographic  | DBN    | School Name                           | Cohort   | Total Cohort | Total Grads - n |
|---|--------------|--------|---------------------------------------|----------|--------------|-----------------|
| 0 | Total Cohort | 01M292 | HENRY STREET SCHOOL FOR INTERNATIONAL | 2003     | 5            | s               |
| 1 | Total Cohort | 01M292 | HENRY STREET SCHOOL FOR INTERNATIONAL | 2004     | 55           | 37              |
| 2 | Total Cohort | 01M292 | HENRY STREET SCHOOL FOR INTERNATIONAL | 2005     | 64           | 43              |
| 3 | Total Cohort | 01M292 | HENRY STREET SCHOOL FOR INTERNATIONAL | 2006     | 78           | 43              |
| 4 | Total Cohort | 01M292 | HENRY STREET SCHOOL FOR INTERNATIONAL | 2006 Aug | 78           | 44              |

The **Demographic** and **Cohort** columns are what prevent **DBN** from being unique in the **graduation** data. A **Cohort** appears to refer to the year the data represents, and the **Demographic** appears to refer to a specific demographic group. In this case, we want to pick data from the most recent Cohort available, which is 2006. We also want data from the full cohort, so we'll only pick rows where **Demographic** is **Total Cohort**.

## 1.9 Condensing the Graduation Data Set

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Filter **graduation**, only selecting rows where the **Cohort** column equals **2006.**
- Filter **graduation**, only selecting rows where the **Demographic** column equals **Total Cohort**.
- Display the first few rows of **data["graduation"]** to verify that everything worked properly.

In [0]:
# put your code here
data["graduation"] = data["graduation"].loc[data["graduation"]["Cohort"]=="2006"]
data["graduation"].head()

Demographic     DBN                            School Name Cohort  \
3   Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2006   
10  Total Cohort  01M448    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   2006   
17  Total Cohort  01M450             EAST SIDE COMMUNITY SCHOOL   2006   
24  Total Cohort  01M509                MARTA VALLE HIGH SCHOOL   2006   
31  Total Cohort  01M515  LOWER EAST SIDE PREPARATORY HIGH SCHO   2006   

    Total Cohort Total Grads - n Total Grads - % of cohort Total Regents - n  \
3             78              43                      55.1                36   
10           124              53                      42.7                42   
17            90              70                      77.8                67   
24            84              47                        56                40   
31           193             105                      54.4                91   

   Total Regents - % of cohort Total Regents - % of grads  \
3                         46.2                       83.7   
10                        33.9                       79.2   
17          74.400000000000006                       95.7   
24                        47.6                       85.1   
31                        47.2                       86.7   

              ...            Regents w/o Advanced - n  \
3             ...                                  36   
10            ...                                  34   
17            ...                                  67   
24            ...                                  23   
31            ...                                  22   

   Regents w/o Advanced - % of cohort Regents w/o Advanced - % of grads  \
3                                46.2                              83.7   
10                               27.4                              64.2   
17                 74.400000000000006                              95.7   
24                               27.4                              48.9   
31                               11.4                                21   

   Local - n Local - % of cohort Local - % of grads Still Enrolled - n  \
3          7                   9               16.3                 16   
10        11                 8.9               20.8                 46   
17         3                 3.3                4.3                 15   
24         7   8.300000000000001               14.9                 25   
31        14                 7.3               13.3                 53   

   Still Enrolled - % of cohort Dropped Out - n Dropped Out - % of cohort  
3                          20.5              11                      14.1  
10                         37.1              20        16.100000000000001  
17                         16.7               5                       5.6  
24                         29.8               5                         6  
31                         27.5              35        18.100000000000001  

[5 rows x 23 columns]

In [0]:
data["graduation"] = data["graduation"][data["graduation"]["Demographic"]=="Total Cohort"]

In [0]:
data["graduation"].head()

Demographic     DBN                            School Name Cohort  \
3   Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2006   
10  Total Cohort  01M448    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   2006   
17  Total Cohort  01M450             EAST SIDE COMMUNITY SCHOOL   2006   
24  Total Cohort  01M509                MARTA VALLE HIGH SCHOOL   2006   
31  Total Cohort  01M515  LOWER EAST SIDE PREPARATORY HIGH SCHO   2006   

    Total Cohort Total Grads - n Total Grads - % of cohort Total Regents - n  \
3             78              43                      55.1                36   
10           124              53                      42.7                42   
17            90              70                      77.8                67   
24            84              47                        56                40   
31           193             105                      54.4                91   

   Total Regents - % of cohort Total Regents - % of grads  \
3                         46.2                       83.7   
10                        33.9                       79.2   
17          74.400000000000006                       95.7   
24                        47.6                       85.1   
31                        47.2                       86.7   

              ...            Regents w/o Advanced - n  \
3             ...                                  36   
10            ...                                  34   
17            ...                                  67   
24            ...                                  23   
31            ...                                  22   

   Regents w/o Advanced - % of cohort Regents w/o Advanced - % of grads  \
3                                46.2                              83.7   
10                               27.4                              64.2   
17                 74.400000000000006                              95.7   
24                               27.4                              48.9   
31                               11.4                                21   

   Local - n Local - % of cohort Local - % of grads Still Enrolled - n  \
3          7                   9               16.3                 16   
10        11                 8.9               20.8                 46   
17         3                 3.3                4.3                 15   
24         7   8.300000000000001               14.9                 25   
31        14                 7.3               13.3                 53   

   Still Enrolled - % of cohort Dropped Out - n Dropped Out - % of cohort  
3                          20.5              11                      14.1  
10                         37.1              20        16.100000000000001  
17                         16.7               5                       5.6  
24                         29.8               5                         6  
31                         27.5              35        18.100000000000001  

[5 rows x 23 columns]

## 1.10 Converting AP Test Scores

We're almost ready to combine all of the data sets. The only remaining thing to do is convert the [Advanced Placement (AP)](https://en.wikipedia.org/wiki/Advanced_Placement_exams) test scores from strings to numeric values. High school students take the AP exams before applying to college. There are several AP exams, each corresponding to a school subject. High school students who earn high scores may receive college credit.

AP exams have a 1 to 5 scale; 3 or higher is a passing score. Many high school students take AP exams -- particularly those who attend academically challenging institutions. AP exams are much more rare in schools that lack funding or academic rigor.

It will be interesting to find out whether AP exam scores are correlated with SAT scores across high schools. To determine this, we'll need to convert the AP exam scores in the **ap_2010** data set to numeric values first.

There are three columns we'll need to convert:

- **AP Test Takers** (note that there's a trailing space in the column name)
- **Total Exams Taken**
- **Number of Exams with scores 3 4 or 5**

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Convert each of the following columns in **ap_2010** to numeric values using the [pandas.to_numeric()](http://pandas.pydata.org/pandas-docs/version/0.17.0/generated/pandas.to_numeric.html) function with the keyword argument **errors="coerce".**
    - **AP Test Takers**
    - **Total Exams Taken**
    - **Number of Exams with scores 3 4 or 5**
- Display the column types using the **dtypes** attribute.

In [0]:
# put your code here

data["ap_2010"]["AP Test Takers "] = pd.to_numeric(data["ap_2010"]["AP Test Takers "], errors="coerce")
np.dtype(data["ap_2010"]["AP Test Takers "])

dtype('float64')

In [0]:
data["ap_2010"]["Total Exams Taken"] = pd.to_numeric(data["ap_2010"]["Total Exams Taken"], errors="coerce")
np.dtype(data["ap_2010"]["Total Exams Taken"])

dtype('float64')

In [0]:
data["ap_2010"]["Number of Exams with scores 3 4 or 5"] = pd.to_numeric(data["ap_2010"]["Number of Exams with scores 3 4 or 5"],  errors="coerce")
np.dtype(data["ap_2010"]["Number of Exams with scores 3 4 or 5"])

dtype('float64')

## 1.11 Left, Right, Inner, and Outer Joins

Before we merge our data, we'll need to decide on the merge strategy we want to use. We'll be using the pandas [pandas.DataFrame.merge()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html) function, which supports four types of joins -- **left**, **right**, **inner**, and **outer**. Each of these join types dictates how pandas combines the rows.

We'll be using the **DBN** column to identify matching rows across data sets. In other words, the values in that column will help us know which row from the first data set to combine with which row in the second data set.

There may be **DBN** values that exist in one data set but not in another. This is partly because the data is from different years. Each data set also has inconsistencies in terms of how it was gathered. Human error (and other types of errors) may also play a role. Therefore, we may not find matches for the **DBN** values in **sat_results** in all of the other data sets, and other data sets may have **DBN** values that don't exist in **sat_results**.

We'll merge two data sets at a time. For example, we'll merge **sat_results** and **hs_directory**, then merge the result with **ap_2010**, then merge the result of that with **class_size**. We'll continue combining data sets in this way until we've merged all of them. Afterwards, we'll have roughly the same number of rows, but each row will have columns from all of the data sets.

The merge strategy we pick will affect the number of rows we end up with. Let's take a look at each strategy.

Let's say we're merging the following two data sets:

<left><img width="300" src="https://drive.google.com/uc?export=view&id=1Vlypix_SIkxCdRS0ABvO4tGiuvFLg321"></left>

With an **inner merge**, we'd only combine rows where the same **DBN** exists in both data sets. We'd end up with this result:

<left><img width="600" src="https://drive.google.com/uc?export=view&id=1LR4c8louX-JAZFYta_Y99FLsGkCf9grr"></left>

With a **left merge**, we'd only use **DBN** values from the dataframe on the "left" of the merge. In this case, **sat_results** is on the left. Some of the DBNs in **sat_results** don't exist in **class_size**, though. The merge will handle this by assiging null values to the columns in **sat_results** that don't have corresponding data in **class_size.**

<left><img width="600" src="https://drive.google.com/uc?export=view&id=1hPoJ5wLECEzz25jrTP5bw9oZ0eerNi2p"></left>

With a **right merge**, we'll only use **DBN** values from the dataframe on the "right" of the merge. In this case, **class_size** is on the right:

<left><img width="600" src="https://drive.google.com/uc?export=view&id=1YYdf4iEMtHYqRBMTEFcfuTyu9zdFlnx7"></left>

With an outer merge, we'll take any DBN values from either sat_results or class_size:

<left><img width="600" src="https://drive.google.com/uc?export=view&id=1sl5wCK3WZ3lTzJm8JUn-bg4MoXl3xSe9"></left>

As you can see, each merge strategy has its advantages. Depending on the strategy we choose, we may preserve rows at the expense of having more missing column data, or minimize missing data at the expense of having fewer rows. Choosing a merge strategy is an important decision; it's worth thinking about your data carefully, and what trade-offs you're willing to make.

Because this project is concerned with determing demographic factors that correlate with SAT score, we'll want to preserve as many rows as possible from **sat_results** while minimizing null values.

This means that we may need to use different merge strategies with different data sets. Some of the data sets have a lot of missing **DBN** values. This makes a **left** join more appropriate, because we don't want to lose too many rows when we merge. If we did an **inner** join, we would lose the data for many high schools.

Some data sets have **DBN** values that are almost identical to those in **sat_results**. Those data sets also have information we need to keep. Most of our analysis would be impossible if a significant number of rows was missing from **demographics**, for example. Therefore, we'll do an inner join to avoid missing data in these columns.

##  1.12 Performing the Left Joins

Both the **ap_2010** and the **graduation** data sets have many missing **DBN** values, so we'll use a left join when we merge the **sat_results** data set with them. Because we're using a **left** join, our final dataframe will have all of the same **DBN** values as the original **sat_results** dataframe.

We'll need to use the pandas **df.merge()** method to merge dataframes. The "left" dataframe is the one we call the method on, and the "right" dataframe is the one we pass into **df.merge()**.

Because we're using the **DBN** column to join the dataframes, we'll need to specify the keyword argument **on="DBN"** when calling **pandas.DataFrame.merge().**

First, we'll assign **data["sat_results"]** to the variable **combined**. Then, we'll merge all of the other dataframes with **combined**. When we're finished, **combined** will have all of the columns from all of the data sets.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Use the pandas [pandas.DataFrame.merge()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html) method to merge the **ap_2010** data set into **combined.**
    - Make sure to specify **how="left"** as a keyword argument to indicate the correct join type.
    - Make sure to assign the result of the merge operation back to **combined.**
- Use the pandas **df.merge()** method to merge the **graduation** data set into **combined.**
    - Make sure to specify **how="left"** as a keyword argument to get the correct join type.
    - Make sure to assign the result of the merge operation back to **combined.**
- Display the first few rows of **combined** to verify that the correct operations occurred.
- Use the [pandas.DataFrame.shape](http://pandas.pydata.org/pandas-docs/version/0.17.0/generated/pandas.DataFrame.shape.html) attribute to display the shape of the dataframe and see how many rows now exist.

In [0]:
# put your code here
combined = data["sat_results"]
combined = combined.merge(data["ap_2010"], on="DBN",how="left")
combined

DBN                                        SCHOOL NAME  \
0    01M292      HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1    01M448                UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2    01M450                         EAST SIDE COMMUNITY SCHOOL   
3    01M458                          FORSYTH SATELLITE ACADEMY   
4    01M509                            MARTA VALLE HIGH SCHOOL   
5    01M515            LOWER EAST SIDE PREPARATORY HIGH SCHOOL   
6    01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND ...   
7    01M650                               CASCADES HIGH SCHOOL   
8    01M696                     BARD HIGH SCHOOL EARLY COLLEGE   
9    02M047  47 THE AMERICAN SIGN LANGUAGE AND ENGLISH SECO...   
10   02M288                       FOOD AND FINANCE HIGH SCHOOL   
11   02M294                               ESSEX STREET ACADEMY   
12   02M296              HIGH SCHOOL OF HOSPITALITY MANAGEMENT   
13   02M298                                   PACE HIGH SCHOOL   
14   02M300  URBAN ASSEMBLY SCHOOL OF DESIGN AND CONSTRUCTI...   
15   02M303                         FACING HISTORY SCHOOL, THE   
16   02M305  URBAN ASSEMBLY ACADEMY OF GOVERNMENT AND LAW, THE   
17   02M308                       LOWER MANHATTAN ARTS ACADEMY   
18   02M313  JAMES BALDWIN SCHOOL, THE: A SCHOOL FOR EXPEDI...   
19   02M316  URBAN ASSEMBLY SCHOOL OF BUSINESS FOR YOUNG WO...   
20   02M374                          GRAMERCY ARTS HIGH SCHOOL   
21   02M376                                        NYC ISCHOOL   
22   02M392                         MANHATTAN BUSINESS ACADEMY   
23   02M393                          BUSINESS OF SPORTS SCHOOL   
24   02M394                           EMMA LAZARUS HIGH SCHOOL   
25   02M399         THE HIGH SCHOOL FOR LANGUAGE AND DIPLOMACY   
26   02M400              HIGH SCHOOL FOR ENVIRONMENTAL STUDIES   
27   02M407              INSTITUTE FOR COLLABORATIVE EDUCATION   
28   02M408           PROFESSIONAL PERFORMING ARTS HIGH SCHOOL   
29   02M411                  BARUCH COLLEGE CAMPUS HIGH SCHOOL   
..      ...                                                ...   
449  30Q580          BACCALAUREATE SCHOOL FOR GLOBAL EDUCATION   
450  31R047          CSI HIGH SCHOOL FOR INTERNATIONAL STUDIES   
451  31R064        GAYNOR MCCOWN EXPEDITIONARY LEARNING SCHOOL   
452  31R080                     THE MICHAEL J. PETRIDES SCHOOL   
453  31R440                               NEW DORP HIGH SCHOOL   
454  31R445                          PORT RICHMOND HIGH SCHOOL   
455  31R450                                 CURTIS HIGH SCHOOL   
456  31R455                            TOTTENVILLE HIGH SCHOOL   
457  31R460                        SUSAN E. WAGNER HIGH SCHOOL   
458  31R470                                CONCORD HIGH SCHOOL   
459  31R600  RALPH R. MCKEE CAREER AND TECHNICAL EDUCATION ...   
460  31R605                STATEN ISLAND TECHNICAL HIGH SCHOOL   
461  31R607    YOUNG ADULT BOROUGH CENTER AT STATEN IS TECH HS   
462  32K403               ACADEMY FOR ENVIRONMENTAL LEADERSHIP   
463  32K545        EBC HIGH SCHOOL FOR PUBLIC SERVICE–BUSHWICK   
464  32K549                 BUSHWICK SCHOOL FOR SOCIAL JUSTICE   
465  32K552                          ACADEMY OF URBAN PLANNING   
466  32K554               ALL CITY LEADERSHIP SECONDARY SCHOOL   
467  32K556  BUSHWICK LEADERS HIGH SCHOOL FOR ACADEMIC EXCE...   
468  32K564                     BUSHWICK COMMUNITY HIGH SCHOOL   
469  75K371                      P.S. 371 s LILLIAN L. RASHKIS   
470  75M035                                           P.S. 035   
471  75Q256                                          P.S. Q256   
472  75Q811                                          P.S. Q811   
473  75R025            SOUTH RICHMOND HIGH SCHOOL I.S./P.S. 25   
474  75X012                   P.S. X012 LEWIS AND CLARK SCHOOL   
475  75X754            J. M. RAPPORT SCHOOL CAREER DEVELOPMENT   
476  79M645         SCHOOL FOR COOPERATIVE TECHNICAL EDUCATION   
477  79Q950                                GED P

In [0]:
combined = combined.merge(data["graduation"], on="DBN",how="left")
combined.head()

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                      7                             414                 401   
4                     44                             390                 433   

  SAT Writing Avg. Score                    SchoolName  AP Test Takers   \
0                    363                           NaN              NaN   
1                    366  UNIVERSITY NEIGHBORHOOD H.S.             39.0   
2                    370        EAST SIDE COMMUNITY HS             19.0   
3                    359                           NaN              NaN   
4                    384                           NaN              NaN   

   Total Exams Taken  Number of Exams with scores 3 4 or 5  \
0                NaN                                   NaN   
1               49.0                                  10.0   
2               21.0                                   NaN   
3                NaN                                   NaN   
4                NaN                                   NaN   

             ...            Regents w/o Advanced - n  \
0            ...                                  36   
1            ...                                  34   
2            ...                                  67   
3            ...                                 NaN   
4            ...                                  23   

  Regents w/o Advanced - % of cohort Regents w/o Advanced - % of grads  \
0                               46.2                              83.7   
1                               27.4                              64.2   
2                 74.400000000000006                              95.7   
3                                NaN                               NaN   
4                               27.4                              48.9   

   Local - n Local - % of cohort Local - % of grads Still Enrolled - n  \
0          7                   9               16.3                 16   
1         11                 8.9               20.8                 46   
2          3                 3.3                4.3                 15   
3        NaN                 NaN                NaN                NaN   
4          7   8.300000000000001               14.9                 25   

  Still Enrolled - % of cohort Dropped Out - n Dropped Out - % of cohort  
0                         20.5              11                      14.1  
1                         37.1              20        16.100000000000001  
2                         16.7               5                       5.6  
3                          NaN             NaN                       NaN  
4                         29.8               5                         6  

[5 rows x 32 columns]

In [0]:
combined.shape

(479, 32)

## 1.13 Performing the Inner Joins

Now that we've performed the left joins, we still have to merge **class_size**, **demographics**, **survey**, and **hs_directory** into **combined**. Because these files contain information that's more valuable to our analysis and also have fewer missing **DBN** values, we'll use the **inner** join type.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Merge **class_size** into **combined**. Then, merge **demographics**, **survey**, and **hs_directory** into **combined** one by one, in that order.
    - Be sure to follow the exact order above.
    - Remember to specify the correct column to join on, as well as the correct join type.
- Display the first few rows of **combined** to verify that the correct operations occurred.
- Call **pandas.DataFrame.shape()** to display the shape of the dataframe to see how many rows now exist.

In [0]:
# put your code here
combined = combined.merge(data["class_size"], on="DBN", how="inner")
combined.head()

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                      7                             414                 401   
4                     44                             390                 433   

  SAT Writing Avg. Score                    SchoolName  AP Test Takers   \
0                    363                           NaN              NaN   
1                    366  UNIVERSITY NEIGHBORHOOD H.S.             39.0   
2                    370        EAST SIDE COMMUNITY HS             19.0   
3                    359                           NaN              NaN   
4                    384                           NaN              NaN   

   Total Exams Taken  Number of Exams with scores 3 4 or 5  \
0                NaN                                   NaN   
1               49.0                                  10.0   
2               21.0                                   NaN   
3                NaN                                   NaN   
4                NaN                                   NaN   

               ...               Still Enrolled - n  \
0              ...                               16   
1              ...                               46   
2              ...                               15   
3              ...                              NaN   
4              ...                               25   

  Still Enrolled - % of cohort Dropped Out - n  Dropped Out - % of cohort  \
0                         20.5              11                       14.1   
1                         37.1              20         16.100000000000001   
2                         16.7               5                        5.6   
3                          NaN             NaN                        NaN   
4                         29.8               5                          6   

  NUMBER OF STUDENTS / SEATS FILLED NUMBER OF SECTIONS AVERAGE CLASS SIZE  \
0                         88.000000           4.000000          22.564286   
1                        105.687500           4.750000          22.231250   
2                         57.600000           2.733333          21.200000   
3                         28.600000           1.200000          23.000000   
4                         69.642857           3.000000          23.571429   

  SIZE OF SMALLEST CLASS SIZE OF LARGEST CLASS SCHOOLWIDE PUPIL-TEACHER RATIO  
0                  18.50             26.571429                            NaN  
1                  18.25             27.062500                            NaN  
2                  19.40             22.866667                            NaN  
3                  22.60             23.400000                            NaN  
4                  20.00             27.357143                            NaN  

[5 rows x 38 columns]

In [0]:
combined = combined.merge(data["demographics"], on="DBN", how="inner")
combined.head()

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                      7                             414                 401   
4                     44                             390                 433   

  SAT Writing Avg. Score                    SchoolName  AP Test Takers   \
0                    363                           NaN              NaN   
1                    366  UNIVERSITY NEIGHBORHOOD H.S.             39.0   
2                    370        EAST SIDE COMMUNITY HS             19.0   
3                    359                           NaN              NaN   
4                    384                           NaN              NaN   

   Total Exams Taken  Number of Exams with scores 3 4 or 5    ...      \
0                NaN                                   NaN    ...       
1               49.0                                  10.0    ...       
2               21.0                                   NaN    ...       
3                NaN                                   NaN    ...       
4                NaN                                   NaN    ...       

  black_num black_per hispanic_num  hispanic_per white_num white_per male_num  \
0       123      29.1          227          53.8         7       1.7    259.0   
1        89      22.6          181          45.9         9       2.3    226.0   
2       143      23.9          331          55.4        62      10.4    327.0   
3        77      34.4          133          59.4         8       3.6     97.0   
4       116      31.6          209          56.9         6       1.6    170.0   

  male_per female_num female_per  
0     61.4      163.0       38.6  
1     57.4      168.0       42.6  
2     54.7      271.0       45.3  
3     43.3      127.0       56.7  
4     46.3      197.0       53.7  

[5 rows x 75 columns]

In [0]:
combined = combined.merge(data["survey"], on="DBN", how="inner")
combined.head()

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                      7                             414                 401   
4                     44                             390                 433   

  SAT Writing Avg. Score                    SchoolName  AP Test Takers   \
0                    363                           NaN              NaN   
1                    366  UNIVERSITY NEIGHBORHOOD H.S.             39.0   
2                    370        EAST SIDE COMMUNITY HS             19.0   
3                    359                           NaN              NaN   
4                    384                           NaN              NaN   

   Total Exams Taken  Number of Exams with scores 3 4 or 5    ...      \
0                NaN                                   NaN    ...       
1               49.0                                  10.0    ...       
2               21.0                                   NaN    ...       
3                NaN                                   NaN    ...       
4                NaN                                   NaN    ...       

  eng_t_11 aca_t_11 saf_s_11  com_s_11 eng_s_11 aca_s_11 saf_tot_11  \
0      6.1      6.5      6.0       5.6      6.1      6.7        6.7   
1      6.6      7.3      6.0       5.7      6.3      7.0        6.8   
2      8.0      8.8      NaN       NaN      NaN      NaN        7.9   
3      8.9      8.9      6.8       6.1      6.1      6.8        7.8   
4      6.1      6.8      6.4       5.9      6.4      7.0        6.9   

  com_tot_11 eng_tot_11 aca_tot_11  
0        6.2        6.6        7.0  
1        6.3        6.7        7.2  
2        7.9        7.9        8.4  
3        7.1        7.2        7.8  
4        6.2        6.6        7.0  

[5 rows x 97 columns]

In [0]:
data["hs_directory"]["DBN"] = data["hs_directory"]["dbn"]
combined = combined.merge(data["hs_directory"], on="DBN", how="inner")
combined.head()

DBN                                        SCHOOL NAME  \
0  01M292      HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448                UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                         EAST SIDE COMMUNITY SCHOOL   
3  01M509                            MARTA VALLE HIGH SCHOOL   
4  01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND ...   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                     44                             390                 433   
4                    159                             522                 574   

  SAT Writing Avg. Score                      SchoolName  AP Test Takers   \
0                    363                             NaN              NaN   
1                    366    UNIVERSITY NEIGHBORHOOD H.S.             39.0   
2                    370          EAST SIDE COMMUNITY HS             19.0   
3                    384                             NaN              NaN   
4                    525  NEW EXPLORATIONS SCI,TECH,MATH            255.0   

   Total Exams Taken  Number of Exams with scores 3 4 or 5  \
0                NaN                                   NaN   
1               49.0                                  10.0   
2               21.0                                   NaN   
3                NaN                                   NaN   
4              377.0                                 191.0   

                         ...                         priority08 priority09  \
0                        ...                                NaN        NaN   
1                        ...                                NaN        NaN   
2                        ...                                NaN        NaN   
3                        ...                                NaN        NaN   
4                        ...                                NaN        NaN   

  priority10                                         Location 1  \
0        NaN  220 Henry Street\nNew York, NY 10002\n(40.7137...   
1        NaN  200 Monroe Street\nNew York, NY 10002\n(40.712...   
2        NaN  420 East 12 Street\nNew York, NY 10009\n(40.72...   
3        NaN  145 Stanton Street\nNew York, NY 10002\n(40.72...   
4        NaN  111 Columbia Street\nNew York, NY 10002\n(40.7...   

  Community Board Council District Census Tract        BIN           BBL  \
0             3.0              1.0        201.0  1003223.0  1.002690e+09   
1             3.0              1.0        202.0  1003214.0  1.002590e+09   
2             3.0              2.0         34.0  1005974.0  1.004390e+09   
3             3.0              1.0       3001.0  1004323.0  1.003540e+09   
4             3.0              2.0       2201.0  1004070.0  1.003350e+09   

                                                 NTA  
0  Lower East Side                               ...  
1  Lower East Side                               ...  
2  East Village                                  ...  
3  Chinatown                                     ...  
4  Lower East Side                               ...  

[5 rows x 161 columns]

In [0]:
combined.shape

(363, 161)

##  1.14 Filling in Missing Values

You may have noticed that the inner joins resulted in 116 fewer rows in **sat_results**. This is because pandas couldn't find the **DBN** values that existed in **sat_results** in the other data sets. While this is worth investigating, we're currently looking for high-level correlations, so we don't need to dive into which **DBNs** are missing.

You may also have noticed that we now have many columns with null (**NaN**) values. This is because we chose to do **left** joins, where some columns may not have had data. The data set also had some missing values to begin with. If we hadn't performed a **left** join, all of the rows with missing data would have been lost in the merge process, which wouldn't have left us with many high schools in our data set.

There are several ways to handle missing data, and we'll cover them in more detail later on. For now, we'll just fill in the missing values with the overall mean for the column, like so:

<left><img width="500" src="https://drive.google.com/uc?export=view&id=1OmhXzMuPrGSmyyugGXpmrRlLznDHxOeT"></left>

In the diagram above, the mean of the first column is (1800 + 1600 + 2200 + 2300) / 4, or 1975, and the mean of the second column is (20 + 30 + 30 + 50) / 4, or 32.5. We replace the missing values with the means of their respective columns, which allows us to proceed with analyses that can't handle missing values (like correlations).

We can fill in missing data in pandas using the [pandas.DataFrame.fillna()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.fillna.html) method. This method will replace any missing values in a dataframe with the values we specify. We can compute the mean of every column using the [pandas.DataFrame.mean()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.mean.html) method. If we pass the results of the **df.mean()** method into the **df.fillna()** method, pandas will fill in the missing values in each column with the mean of that column.

Here's an example of how we would accomplish this:

```python
means = df.mean()
df = df.fillna(means)
```

Note that if a column consists entirely of null or **NaN** values, pandas won't be able to fill in the missing values when we use the **df.fillna()** method along with the **df.mean()** method, because there won't be a mean.

We should fill any **NaN** or null values that remain after the initial replacement with the value 0. We can do this by passing 0 into the **df.fillna()** method.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Calculate the means of all of the columns in **combined** using the **pandas.DataFrame.mean()** method.
- Fill in any missing values in **combined** with the means of the respective columns using the **pandas.DataFrame.fillna()** method.
- Fill in any remaining missing values in **combined** with 0 using the **df.fillna()** method.
- Display the first few rows of **combined** to verify that the correct operations occurred.

In [0]:
# put your code here
means = combined.mean()
means

AP Test Takers                          1.290288e+02
Total Exams Taken                       1.970385e+02
Number of Exams with scores 3 4 or 5    1.534500e+02
Total Cohort                            1.938710e+02
NUMBER OF STUDENTS / SEATS FILLED       1.717624e+02
NUMBER OF SECTIONS                      6.491468e+00
AVERAGE CLASS SIZE                      2.479408e+01
SIZE OF SMALLEST CLASS                  1.984065e+01
SIZE OF LARGEST CLASS                   2.871611e+01
SCHOOLWIDE PUPIL-TEACHER RATIO                   NaN
schoolyear                              2.011201e+07
frl_percent                             6.611901e+01
total_enrollment                        7.930358e+02
ell_num                                 9.205234e+01
ell_percent                             1.333499e+01
sped_num                                1.021350e+02
sped_percent                            1.424766e+01
asian_num                               1.357934e+02
asian_per                               1.0149

In [0]:
combined = combined.fillna(means)

In [0]:
combined = combined.fillna(0)

In [0]:
combined.head()

DBN                                        SCHOOL NAME  \
0  01M292      HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448                UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                         EAST SIDE COMMUNITY SCHOOL   
3  01M509                            MARTA VALLE HIGH SCHOOL   
4  01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND ...   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                     44                             390                 433   
4                    159                             522                 574   

  SAT Writing Avg. Score                      SchoolName  AP Test Takers   \
0                    363                               0       129.028846   
1                    366    UNIVERSITY NEIGHBORHOOD H.S.        39.000000   
2                    370          EAST SIDE COMMUNITY HS        19.000000   
3                    384                               0       129.028846   
4                    525  NEW EXPLORATIONS SCI,TECH,MATH       255.000000   

   Total Exams Taken  Number of Exams with scores 3 4 or 5  \
0         197.038462                                153.45   
1          49.000000                                 10.00   
2          21.000000                                153.45   
3         197.038462                                153.45   
4         377.000000                                191.00   

                         ...                         priority08 priority09  \
0                        ...                                  0          0   
1                        ...                                  0          0   
2                        ...                                  0          0   
3                        ...                                  0          0   
4                        ...                                  0          0   

  priority10                                         Location 1  \
0          0  220 Henry Street\nNew York, NY 10002\n(40.7137...   
1          0  200 Monroe Street\nNew York, NY 10002\n(40.712...   
2          0  420 East 12 Street\nNew York, NY 10009\n(40.72...   
3          0  145 Stanton Street\nNew York, NY 10002\n(40.72...   
4          0  111 Columbia Street\nNew York, NY 10002\n(40.7...   

  Community Board Council District Census Tract        BIN           BBL  \
0             3.0              1.0        201.0  1003223.0  1.002690e+09   
1             3.0              1.0        202.0  1003214.0  1.002590e+09   
2             3.0              2.0         34.0  1005974.0  1.004390e+09   
3             3.0              1.0       3001.0  1004323.0  1.003540e+09   
4             3.0              2.0       2201.0  1004070.0  1.003350e+09   

                                                 NTA  
0  Lower East Side                               ...  
1  Lower East Side                               ...  
2  East Village                                  ...  
3  Chinatown                                     ...  
4  Lower East Side                               ...  

[5 rows x 161 columns]

## 1.15 Adding a School District Column for Mapping

We've finished cleaning and combining our data! We now have a clean data set on which we can base our analysis. Mapping the statistics out on a school district level might be an interesting way to analyze them. Adding a column to the data set that specifies the school district will help us accomplish this.

The school district is just the first two characters of the **DBN**. We can apply a function over the **DBN** column of **combined** that pulls out the first two letters.

For example, we can use indexing to extract the first few characters of a string, like this:

```python
name = "Sinbad"
print(name[0:2])
```

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Write a function that extracts the first two characters of a string and returns them.
- Apply the function to the **DBN** column of **combined**, and assign the result to the **school_dist** column of **combined**.
- Display the first few items in the **school_dist** column of **combined** to verify the results.


In [0]:
# put your code here
def two_c (s):
  return s[:2]

combined["school_dist"] = combined["DBN"].apply(two_c)
combined["school_dist"].head()

0    01
1    01
2    01
3    01
4    01
Name: school_dist, dtype: object

## 1.16 Next Steps

We now have a clean data set we can analyze! We've done a lot in this mission. We've gone from having several messy sources to one clean, combined, data set that's ready for analysis.

Along the way, we've learned about:

- How to handle missing values
- Different types of merges
- How to condense data sets
- How to compute averages across dataframes

Data scientists rarely start out with tidy data sets, which makes cleaning and combining them one of the most critical skills any data professional can learn.
